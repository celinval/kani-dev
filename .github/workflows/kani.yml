# Copyright Kani Contributors
# SPDX-License-Identifier: Apache-2.0 OR MIT
name: Kani CI
on:
  pull_request:
  push:
    # Not just any push, as that includes tags.
    # We don't want to re-trigger this workflow when tagging an existing commit.
    branches:
      - '**'
  workflow_dispatch:
    inputs:
      logLevel:
        description: 'Log level'
        required: false
        default: 'warn'
env:
  RUST_BACKTRACE: 1
  KANI_LOG: ${{ github.event.inputs.logLevel }}

jobs:
  regression:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [macos-11, ubuntu-18.04, ubuntu-20.04, ubuntu-22.04]
    steps:
      - name: Checkout Kani
        uses: actions/checkout@v3

      - name: Setup Kani Dependencies
        uses: ./.github/actions/setup
        with:
          os: ${{ matrix.os }}

      - name: Build Kani
        run: cargo build-dev

      - name: Execute Kani regression
        run: ./scripts/run-dep.sh
