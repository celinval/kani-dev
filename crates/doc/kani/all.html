<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-405f8b29f52305f8.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (506f22b46 2024-09-19)" data-channel="nightly" data-search-js="search-0cfde64e8ad3a7fe.js" data-settings-js="settings-7e3bb6c46e92e77c.js" ><script src="../static.files/storage-29b1d5a9048d38fe.js"></script><script defer src="../static.files/main-59a36e46ca746814.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-40f72c9259523cb3.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../kani/index.html">kani</a><span class="version">0.55.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h3><a href="#macros">Crate Items</a></h3><ul class="block"><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li><li><a href="#derives" title="Derive Macros">Derive Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="futures/struct.JoinHandle.html">futures::JoinHandle</a></li><li><a href="futures/struct.RoundRobin.html">futures::RoundRobin</a></li><li><a href="shadow/struct.ShadowMem.html">shadow::ShadowMem</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="futures/enum.SchedulingAssumption.html">futures::SchedulingAssumption</a></li></ul><h3 id="traits">Traits</h3><ul class="all-items"><li><a href="trait.Arbitrary.html">Arbitrary</a></li><li><a href="futures/trait.SchedulingStrategy.html">futures::SchedulingStrategy</a></li><li><a href="invariant/trait.Invariant.html">invariant::Invariant</a></li></ul><h3 id="macros">Macros</h3><ul class="all-items"><li><a href="macro.arbitrary_tuple.html">arbitrary_tuple</a></li><li><a href="macro.cover.html">cover</a></li><li><a href="macro.generate_arbitrary.html">generate_arbitrary</a></li><li><a href="macro.implies.html">implies</a></li><li><a href="macro.kani_intrinsics.html">kani_intrinsics</a></li><li><a href="macro.kani_lib.html">kani_lib</a></li><li><a href="macro.kani_mem.html">kani_mem</a></li><li><a href="macro.kani_mem_init.html">kani_mem_init</a></li></ul><h3 id="attributes">Attribute Macros</h3><ul class="all-items"><li><a href="contracts/attr.ensures.html">contracts::ensures</a></li><li><a href="contracts/attr.modifies.html">contracts::modifies</a></li><li><a href="contracts/attr.proof_for_contract.html">contracts::proof_for_contract</a></li><li><a href="contracts/attr.requires.html">contracts::requires</a></li><li><a href="contracts/attr.stub_verified.html">contracts::stub_verified</a></li><li><a href="attr.ensures.html">ensures</a></li><li><a href="attr.modifies.html">modifies</a></li><li><a href="attr.proof.html">proof</a></li><li><a href="attr.proof_for_contract.html">proof_for_contract</a></li><li><a href="attr.recursion.html">recursion</a></li><li><a href="attr.requires.html">requires</a></li><li><a href="attr.should_panic.html">should_panic</a></li><li><a href="attr.solver.html">solver</a></li><li><a href="attr.stub.html">stub</a></li><li><a href="attr.stub_verified.html">stub_verified</a></li><li><a href="attr.unwind.html">unwind</a></li></ul><h3 id="derives">Derive Macros</h3><ul class="all-items"><li><a href="derive.Arbitrary.html">Arbitrary</a></li><li><a href="derive.Invariant.html">Invariant</a></li></ul><h3 id="functions">Functions</h3><ul class="all-items"><li><a href="fn.any.html">any</a></li><li><a href="fn.any_where.html">any_where</a></li><li><a href="fn.assert.html">assert</a></li><li><a href="fn.assume.html">assume</a></li><li><a href="fn.check.html">check</a></li><li><a href="fn.concrete_playback_run.html">concrete_playback_run</a></li><li><a href="fn.cover.html">cover</a></li><li><a href="futures/fn.block_on.html">futures::block_on</a></li><li><a href="futures/fn.block_on_with_spawn.html">futures::block_on_with_spawn</a></li><li><a href="futures/fn.spawn.html">futures::spawn</a></li><li><a href="futures/fn.yield_now.html">futures::yield_now</a></li><li><a href="mem/fn.can_dereference.html">mem::can_dereference</a></li><li><a href="mem/fn.can_read_unaligned.html">mem::can_read_unaligned</a></li><li><a href="mem/fn.can_write.html">mem::can_write</a></li><li><a href="mem/fn.can_write_unaligned.html">mem::can_write_unaligned</a></li><li><a href="slice/fn.any_slice_of_array.html">slice::any_slice_of_array</a></li><li><a href="slice/fn.any_slice_of_array_mut.html">slice::any_slice_of_array_mut</a></li><li><a href="vec/fn.any_vec.html">vec::any_vec</a></li><li><a href="vec/fn.exact_vec.html">vec::exact_vec</a></li></ul></section></div></main></body></html>