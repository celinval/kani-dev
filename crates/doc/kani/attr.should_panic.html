<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Specifies that a proof harness is expected to panic.**"><title>should_panic in kani - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-405f8b29f52305f8.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (506f22b46 2024-09-19)" data-channel="nightly" data-search-js="search-0cfde64e8ad3a7fe.js" data-settings-js="settings-7e3bb6c46e92e77c.js" ><script src="../static.files/storage-29b1d5a9048d38fe.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-59a36e46ca746814.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-40f72c9259523cb3.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../kani/index.html">kani</a><span class="version">0.55.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">should_<wbr>panic</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#limitations" title="Limitations">Limitations</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate kani</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Attribute Macro <a href="index.html">kani</a>::<wbr><a class="attr" href="#">should_panic</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>#[should_panic]</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Specifies that a proof harness is expected to panic.**</p>
<p>This attribute allows users to exercise <em>negative verification</em>.
It’s analogous to how
<a href="https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html#testing-panics"><code>#[should_panic]</code></a>
allows users to exercise <a href="https://en.wikipedia.org/wiki/Negative_testing">negative testing</a>
for Rust unit tests.</p>
<h2 id="limitations"><a class="doc-anchor" href="#limitations">§</a>Limitations</h2>
<p>The <code>#[kani::should_panic]</code> attribute verifies that there are one or more failed checks related to panics.
At the moment, it’s not possible to pin it down to specific panics.</p>
</div></details></section></div></main></body></html>