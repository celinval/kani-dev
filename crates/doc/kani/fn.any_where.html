<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This creates a symbolic valid value of type `T`. The value is constrained to be a value accepted by the predicate passed to the filter. You can assign the return value of this function to a variable that you want to make symbolic."><title>any_where in kani - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-405f8b29f52305f8.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (506f22b46 2024-09-19)" data-channel="nightly" data-search-js="search-0cfde64e8ad3a7fe.js" data-settings-js="settings-7e3bb6c46e92e77c.js" ><script src="../static.files/storage-29b1d5a9048d38fe.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-59a36e46ca746814.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-40f72c9259523cb3.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc fn"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../kani/index.html">kani</a><span class="version">0.55.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">any_<wbr>where</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#example" title="Example:">Example:</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate kani</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Function <a href="index.html">kani</a>::<wbr><a class="fn" href="#">any_where</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/kani/lib.rs.html#53">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>pub fn any_where&lt;T: <a class="trait" href="trait.Arbitrary.html" title="trait kani::Arbitrary">Arbitrary</a>, F: <a class="trait" href="https://doc.rust-lang.org/nightly/core/ops/function/trait.FnOnce.html" title="trait core::ops::function::FnOnce">FnOnce</a>(<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.reference.html">&amp;T</a>) -&gt; <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.bool.html">bool</a>&gt;(f: F) -&gt; T</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This creates a symbolic <em>valid</em> value of type <code>T</code>.
The value is constrained to be a value accepted by the predicate passed to the filter.
You can assign the return value of this function to a variable that you want to make symbolic.</p>
<h2 id="example"><a class="doc-anchor" href="#example">§</a>Example:</h2>
<p>In the snippet below, we are verifying the behavior of the function <code>fn_under_verification</code>
under all possible <code>u8</code> input values between 0 and 12.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">let </span>inputA: u8 = kani::any_where(|x| <span class="kw-2">*</span>x &lt; <span class="number">12</span>);
fn_under_verification(inputA);</code></pre></div>
<p>Note: This is a safe construct and can only be used with types that implement the <code>Arbitrary</code>
trait. The Arbitrary trait is used to build a symbolic value that represents all possible
valid values for type <code>T</code>.</p>
</div></details></section></div></main></body></html>