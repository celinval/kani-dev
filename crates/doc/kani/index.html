<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `kani` crate."><title>kani - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-405f8b29f52305f8.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="kani" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (506f22b46 2024-09-19)" data-channel="nightly" data-search-js="search-0cfde64e8ad3a7fe.js" data-settings-js="settings-7e3bb6c46e92e77c.js" ><script src="../static.files/storage-29b1d5a9048d38fe.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-59a36e46ca746814.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-40f72c9259523cb3.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../kani/index.html">kani</a><span class="version">0.55.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#reexports">Crate Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#attributes" title="Attribute Macros">Attribute Macros</a></li><li><a href="#derives" title="Derive Macros">Derive Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">kani</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/kani/lib.rs.html#5-89">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.Invariant"><code>pub use invariant::<a class="trait" href="invariant/trait.Invariant.html" title="trait kani::invariant::Invariant">Invariant</a>;</code></div></li><li><div class="item-name" id="reexport.block_on"><code>pub use futures::<a class="fn" href="futures/fn.block_on.html" title="fn kani::futures::block_on">block_on</a>;</code></div></li><li><div class="item-name" id="reexport.block_on_with_spawn"><code>pub use futures::<a class="fn" href="futures/fn.block_on_with_spawn.html" title="fn kani::futures::block_on_with_spawn">block_on_with_spawn</a>;</code></div></li><li><div class="item-name" id="reexport.spawn"><code>pub use futures::<a class="fn" href="futures/fn.spawn.html" title="fn kani::futures::spawn">spawn</a>;</code></div></li><li><div class="item-name" id="reexport.yield_now"><code>pub use futures::<a class="fn" href="futures/fn.yield_now.html" title="fn kani::futures::yield_now">yield_now</a>;</code></div></li><li><div class="item-name" id="reexport.RoundRobin"><code>pub use futures::<a class="struct" href="futures/struct.RoundRobin.html" title="struct kani::futures::RoundRobin">RoundRobin</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="arbitrary/index.html" title="mod kani::arbitrary">arbitrary</a></div><div class="desc docblock-short">This module introduces the <code>Arbitrary</code> trait as well as implementation for
primitive types and other std containers.</div></li><li><div class="item-name"><a class="mod" href="contracts/index.html" title="mod kani::contracts">contracts</a></div><div class="desc docblock-short">Kani implementation of function contracts.</div></li><li><div class="item-name"><a class="mod" href="futures/index.html" title="mod kani::futures">futures</a></div><div class="desc docblock-short">This module contains functions to work with futures (and async/.await) in Kani.</div></li><li><div class="item-name"><a class="mod" href="invariant/index.html" title="mod kani::invariant">invariant</a></div><div class="desc docblock-short">This module introduces the <code>Invariant</code> trait as well as its implementation
for primitive types.</div></li><li><div class="item-name"><a class="mod" href="mem/index.html" title="mod kani::mem">mem</a></div></li><li><div class="item-name"><a class="mod" href="shadow/index.html" title="mod kani::shadow">shadow</a></div><div class="desc docblock-short">This module contains an API for shadow memory.
Shadow memory is a mechanism by which we can store metadata on memory
locations, e.g. whether a memory location is initialized.</div></li><li><div class="item-name"><a class="mod" href="slice/index.html" title="mod kani::slice">slice</a></div></li><li><div class="item-name"><a class="mod" href="vec/index.html" title="mod kani::vec">vec</a></div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.arbitrary_tuple.html" title="macro kani::arbitrary_tuple">arbitrary_<wbr>tuple</a></div><div class="desc docblock-short">This macro implements <code>kani::Arbitrary</code> on a tuple whose elements
already implement <code>kani::Arbitrary</code> by running <code>kani::any()</code> on
each index of the tuple.</div></li><li><div class="item-name"><a class="macro" href="macro.cover.html" title="macro kani::cover">cover</a></div></li><li><div class="item-name"><a class="macro" href="macro.generate_arbitrary.html" title="macro kani::generate_arbitrary">generate_<wbr>arbitrary</a></div></li><li><div class="item-name"><a class="macro" href="macro.implies.html" title="macro kani::implies">implies</a></div><div class="desc docblock-short"><code>implies!(premise =&gt; conclusion)</code> means that if the <code>premise</code> is true, so
must be the <code>conclusion</code>.</div></li><li><div class="item-name"><a class="macro" href="macro.kani_intrinsics.html" title="macro kani::kani_intrinsics">kani_<wbr>intrinsics</a></div><div class="desc docblock-short">Kani intrinsics contains the public APIs used by users to verify their harnesses.
This macro is a part of kani_core as that allows us to verify even libraries that are no_core
such as core in rust’s std library itself.</div></li><li><div class="item-name"><a class="macro" href="macro.kani_lib.html" title="macro kani::kani_lib">kani_<wbr>lib</a></div><div class="desc docblock-short">Users should only need to invoke this.</div></li><li><div class="item-name"><a class="macro" href="macro.kani_mem.html" title="macro kani::kani_mem">kani_<wbr>mem</a></div></li><li><div class="item-name"><a class="macro" href="macro.kani_mem_init.html" title="macro kani::kani_mem_init">kani_<wbr>mem_<wbr>init</a></div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Arbitrary.html" title="trait kani::Arbitrary">Arbitrary</a></div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.any.html" title="fn kani::any">any</a></div><div class="desc docblock-short">This creates an symbolic <em>valid</em> value of type <code>T</code>. You can assign the return value of this
function to a variable that you want to make symbolic.</div></li><li><div class="item-name"><a class="fn" href="fn.any_where.html" title="fn kani::any_where">any_<wbr>where</a></div><div class="desc docblock-short">This creates a symbolic <em>valid</em> value of type <code>T</code>.
The value is constrained to be a value accepted by the predicate passed to the filter.
You can assign the return value of this function to a variable that you want to make symbolic.</div></li><li><div class="item-name"><a class="fn" href="fn.assert.html" title="fn kani::assert">assert</a></div><div class="desc docblock-short">Creates an assertion of the specified condition and message.</div></li><li><div class="item-name"><a class="fn" href="fn.assume.html" title="fn kani::assume">assume</a></div><div class="desc docblock-short">Creates an assumption that will be valid after this statement run. Note that the assumption
will only be applied for paths that follow the assumption. If the assumption doesn’t hold, the
program will exit successfully.</div></li><li><div class="item-name"><a class="fn" href="fn.check.html" title="fn kani::check">check</a></div><div class="desc docblock-short">Creates an assertion of the specified condition and message, but does not assume it afterwards.</div></li><li><div class="item-name"><a class="fn" href="fn.concrete_playback_run.html" title="fn kani::concrete_playback_run">concrete_<wbr>playback_<wbr>run</a></div><div class="desc docblock-short">NOP <code>concrete_playback</code> for type checking during verification mode.</div></li><li><div class="item-name"><a class="fn" href="fn.cover.html" title="fn kani::cover">cover</a></div><div class="desc docblock-short">Creates a cover property with the specified condition and message.</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.ensures.html" title="attr kani::ensures">ensures</a></div><div class="desc docblock-short">Add a postcondition to this function.</div></li><li><div class="item-name"><a class="attr" href="attr.modifies.html" title="attr kani::modifies">modifies</a></div><div class="desc docblock-short">Declaration of an explicit write-set for the annotated function.</div></li><li><div class="item-name"><a class="attr" href="attr.proof.html" title="attr kani::proof">proof</a></div><div class="desc docblock-short">Marks a Kani proof harness</div></li><li><div class="item-name"><a class="attr" href="attr.proof_for_contract.html" title="attr kani::proof_for_contract">proof_<wbr>for_<wbr>contract</a></div><div class="desc docblock-short">Designates this function as a harness to check a function contract.</div></li><li><div class="item-name"><a class="attr" href="attr.recursion.html" title="attr kani::recursion">recursion</a></div><div class="desc docblock-short">Specifies that a function contains recursion for contract instrumentation.**</div></li><li><div class="item-name"><a class="attr" href="attr.requires.html" title="attr kani::requires">requires</a></div><div class="desc docblock-short">Add a precondition to this function.</div></li><li><div class="item-name"><a class="attr" href="attr.should_panic.html" title="attr kani::should_panic">should_<wbr>panic</a></div><div class="desc docblock-short">Specifies that a proof harness is expected to panic.**</div></li><li><div class="item-name"><a class="attr" href="attr.solver.html" title="attr kani::solver">solver</a></div><div class="desc docblock-short">Select the SAT solver to use with CBMC for this harness</div></li><li><div class="item-name"><a class="attr" href="attr.stub.html" title="attr kani::stub">stub</a></div><div class="desc docblock-short">Specify a function/method stub pair to use for proof harness</div></li><li><div class="item-name"><a class="attr" href="attr.stub_verified.html" title="attr kani::stub_verified">stub_<wbr>verified</a></div><div class="desc docblock-short"><code>stub_verified(TARGET)</code> is a harness attribute (to be used on
<a href="attr.proof.html" title="attr kani::proof"><code>proof</code></a> or <a href="attr.proof_for_contract.html" title="attr kani::proof_for_contract"><code>proof_for_contract</code></a>
function) that replaces all occurrences of <code>TARGET</code> reachable from this
harness with a stub generated from the contract on <code>TARGET</code>.</div></li><li><div class="item-name"><a class="attr" href="attr.unwind.html" title="attr kani::unwind">unwind</a></div><div class="desc docblock-short">Set Loop unwind limit for proof harnesses
The attribute <code>#[kani::unwind(arg)]</code> can only be called alongside <code>#[kani::proof]</code>.
arg - Takes in a integer value (u32) that represents the unwind value for the harness.</div></li></ul><h2 id="derives" class="section-header">Derive Macros<a href="#derives" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="derive" href="derive.Arbitrary.html" title="derive kani::Arbitrary">Arbitrary</a></div><div class="desc docblock-short">Allow users to auto generate <code>Arbitrary</code> implementations by using
<code>#[derive(Arbitrary)]</code> macro.</div></li><li><div class="item-name"><a class="derive" href="derive.Invariant.html" title="derive kani::Invariant">Invariant</a></div><div class="desc docblock-short">Allow users to auto generate <code>Invariant</code> implementations by using
<code>#[derive(Invariant)]</code> macro.</div></li></ul></section></div></main></body></html>