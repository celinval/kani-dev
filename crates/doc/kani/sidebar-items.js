window.SIDEBAR_ITEMS = {"attr":["ensures","modifies","proof","proof_for_contract","recursion","requires","should_panic","solver","stub","stub_verified","unwind"],"derive":["Arbitrary","Invariant"],"fn":["any","any_where","assert","assume","check","concrete_playback_run","cover"],"macro":["arbitrary_tuple","cover","generate_arbitrary","implies","kani_intrinsics","kani_lib","kani_mem","kani_mem_init"],"mod":["arbitrary","contracts","futures","invariant","mem","shadow","slice","vec"],"trait":["Arbitrary"]};