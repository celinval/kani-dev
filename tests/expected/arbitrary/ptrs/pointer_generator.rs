// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
//
// kani-flags: -Z mem-predicates
//! Check the behavior of the new `PointerGenerator`.
extern crate kani;

use kani::{cover, AllocationStatus, PointerGenerator};

/// Harness that checks that all cases are covered and the code behaves as expected.
///
/// Note that for `DeadObj`, `Dangling`, and `OutBounds` the predicate will fail due to demonic non-determinism.
#[kani::proof]
fn check_arbitrary_ptr() {
    let mut generator = PointerGenerator::<char, 3>::new();
    let arbitrary = generator.any_alloc_status();
    let ptr = arbitrary.ptr;
    match arbitrary.status {
        AllocationStatus::Dangling => {
            cover!(true, "Dangling");
            assert!(!kani::mem::can_write_unaligned(ptr), "Dangling write");
        }
        AllocationStatus::Null => {
            assert!(!kani::mem::can_write_unaligned(ptr), "NullPtr");
        }
        AllocationStatus::DeadObj => {
            // Due to demonic non-determinism, the API will trigger an error.
            assert!(!kani::mem::can_write_unaligned(ptr), "DeadObj");
        }
        AllocationStatus::OutBounds => {
            assert!(!kani::mem::can_write_unaligned(ptr), "OutBounds");
        }
        AllocationStatus::InBounds => {
            // This should always succeed
            assert!(kani::mem::can_write_unaligned(ptr), "InBounds");
        }
    };
}
