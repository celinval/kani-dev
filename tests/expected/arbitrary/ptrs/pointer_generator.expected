Checking harness check_arbitrary_ptr...

Status: SUCCESS\
Description: ""OutBounds""

Status: SUCCESS\
Description: ""InBounds""

Status: SUCCESS\
Description: ""NullPtr""

Status: FAILURE\
Description: ""DeadObj""

Status: SATISFIED\
Description: "Dangling"

Status: UNREACHABLE\
Description: ""Dangling write""

Verification failed for - check_arbitrary_ptr
